### DISCUSSION TITLE
Extruder heating immediately instead of waiting for bed temp to stabilize

### DISCUSSION THREAD

**Discussion by Lincoln-Jones** [jschuh/klipper-macros#191] [Q&A] [answered]
I've been using klipper-macros successfully with Orcaslicer and a Tronxy X5SA I converted to Klipper for a while, but I just finished building and configuring a VzBot 330 and I'm having an issue. When the print starts, the extruder starts heating immediately instead of waiting for the bed to heat up and stabilize. Which causes lots of oozing and means I have to preheat the bed manually before starting the print.

I've attached a [zip of config files](https://github.com/jschuh/klipper-macros/files/13223397/config.zip), but here are some relevant config setting snippets (with commented out stuff removed for brevity)

**_km_options**

```
[gcode_macro _km_options]
variable_start_purge_length: 30 # This value works for most setups.
variable_start_extruder_preheat_scale: 0.0
variable_start_end_park_y: 315
variable_start_purge_clearance: 10
variable_start_purge_prime_length: 7
```

**OrcaSlicer Machine Start Gcode**

```
M190 S0
M109 S0
_PRINT_START_PHASE_INIT EXTRUDER={first_layer_temperature[initial_tool]} BED=[first_layer_bed_temperature] MESH_MIN={first_layer_print_min[0]},{first_layer_print_min[1]} MESH_MAX={first_layer_print_max[0]},{first_layer_print_max[1]} LAYERS={total_layer_count} NOZZLE_SIZE={nozzle_diameter[0]}
; Insert custom gcode here.
_PRINT_START_PHASE_PREHEAT
; Insert custom gcode here.
_PRINT_START_PHASE_PROBING
; Insert custom gcode here.
_PRINT_START_PHASE_EXTRUDER
; Insert custom gcode here.
_PRINT_START_PHASE_PURGE
```

I've tried a few different things to fix it. Tried removing the M190/M109 lines. Tried removing the phases from the slicer start gcode and simply using `PRINT_START`. Tried rearranging the extruder & bed values in the init phase. Tried changing the name of the values Orcaslicer is passing. Tried adjusting the `variable_start_extruder_preheat_scale` value to 0.2 or 0.5 to see if it changes the initial target temp, but it doesn't.

Here's what the klipper log looks like when the print starts:

```
Stats 50149.5: gcodein=0  mcu: mcu_awake=0.000 mcu_task_avg=0.000002 mcu_task_stddev=0.000001 bytes_write=21831981 bytes_read=4751979 bytes_retransmit=9 bytes_invalid=0 send_seq=375119 receive_seq=375119 retransmit_seq=2 srtt=0.000 rttvar=0.000 rto=0.025 ready_bytes=0 upcoming_bytes=0 freq=400000131  FLY-Super8: temp=24.0 raspberry_pi: temp=32.1 heater_bed: target=0 temp=29.1 pwm=0.000 chamber_temp: temp=21.6 water_loop: temp=21.1 sysload=1.20 cputime=1521.917 memavail=3406448 print_time=9395.144 buffer_time=0.000 print_stall=0 extruder: target=0 temp=23.5 pwm=0.000
Starting SD card print (position 0)
Stats 50150.5: gcodein=0  mcu: mcu_awake=0.000 mcu_task_avg=0.000002 mcu_task_stddev=0.000001 bytes_write=21832033 bytes_read=4752256 bytes_retransmit=9 bytes_invalid=0 send_seq=375121 receive_seq=375121 retransmit_seq=2 srtt=0.000 rttvar=0.000 rto=0.025 ready_bytes=0 upcoming_bytes=0 freq=400000130 sd_pos=40849 FLY-Super8: temp=24.2 raspberry_pi: temp=32.1 heater_bed: target=57 temp=29.0 pwm=0.000 chamber_temp: temp=21.7 water_loop: temp=21.1 sysload=1.20 cputime=1522.021 memavail=3405976 print_time=9450.066 buffer_time=0.000 print_stall=0 extruder: target=210 temp=23.5 pwm=0.000
Heater heater_bed approaching new target of 57.000
Heater extruder approaching new target of 210.000
```

Included the stats to show that the extruder target goes from 0 before "starting SD card print" to 210 after.

Also here's what the start of the gcode file looks like:

```
M73 P0 R42
;TYPE:Custom
M190 S0 ; Remove this if autoemit_temperature_commands is off in Prusa Slicer 2.6 and later
M109 S0 ; Remove this if autoemit_temperature_commands is off in Prusa Slicer 2.6 and later
_PRINT_START_PHASE_INIT EXTRUDER=210 BED=55 MESH_MIN=153.141,155.456 MESH_MAX=178.177,177.08 LAYERS=200 NOZZLE_SIZE=0.4
; Insert custom gcode here.
_PRINT_START_PHASE_PREHEAT
; Insert custom gcode here.
_PRINT_START_PHASE_PROBING
; Insert custom gcode here.
_PRINT_START_PHASE_EXTRUDER
; Insert custom gcode here.
_PRINT_START_PHASE_PURGE

; This is the place to put slicer purge lines if you haven't set a non-zero
; variable_start_purge_length to have START_PRINT automatically calculate and 
; perform the purge (e.g. if using a Mosaic Palette, which requires the slicer
; to generate the purge).
G90
G21
M83 ; use relative distances for extrusion
; filament start gcode
M106 S0
;LAYER_CHANGE
;Z:0.25
;HEIGHT:0.25
;BEFORE_LAYER_CHANGE
;0.25
BEFORE_LAYER_CHANGE HEIGHT=0.25 LAYER=0
G10 ; retract
;AFTER_LAYER_CHANGE
;0.25
AFTER_LAYER_CHANGE
;_SET_FAN_SPEED_CHANGING_LAYER
SET_VELOCITY_LIMIT ACCEL=1500 ACCEL_TO_DECEL=750
```

I've checked to make sure there are no conflicting macros, and there don't seem to be any. Only ones I could find were in mainsail.cfg, but commenting out that include didn't fix the issue.

I just installed klipper-macros today, so it should be up-to-date. My X5SA has an older version of klipper-macros on it, because I stopped updating stuff on that printer to prevent having to do work on it while also building this VzBot. (and I needed the X5SA to keep printing parts for the VzBot reliably) I first tried copying the _km_options section over from the X5SA, but it complained that `variable_start_extruder_set_target_before_level` was deprecated, so I started from scratch and used the X5SA as a reference.

Also the other main difference between these printers is the VzBot has no bed probe. (yet) Should I be removing the "probing" state from the slicer start gcode?

I don't have `bed_mesh` or a virtual z endstop configured, so I wouldn't think `variable_start_level_bed_at_temp` or `variable_start_extruder_probing_temp` would affect anything. I would think setting `variable_start_extruder_preheat_scale` to 0.0 would be all I'd need.

Anyway, long post, but I'm really hoping to get this figured out. Anyone see anything obviously misconfigured, or have suggestions of things I should try? Is there any chance this is a bug?

**Comment by jschuh** [comment, marked as answer]
First, thanks for the amazingly thorough troubleshooting request. Unfortunately, the macros are working as intended in this case. If there's no probing or other operations, the extruder and the bed are heated simultaneously to their target temps.

I never really thought about doing it differently because I don't want to wait on the bed first stabilizing if there's no probing, etc, and I rely on the purge movements to deal with the nozzle ooze.

I'm not sure I want to change this, but I'll think about it.

**Reply by Lincoln-Jones** [reply]
Ah ok, I guess that makes sense. I'll leave my reasons why I prefer waiting to heat the hotend as some food for thought:

- The Goliath hotend on the VzBot has a _looooong_ melt zone, so as a result it has _a lot_ of melted plastic available to ooze out.
- The Goliath hotend also draws a lot of amps and heats quickly, so there isn't much of a wait for it to come to temp.
- When I've tried to let the purge moves take care of the ooze, that oozed filament often sticks to the nozzle and rides over to the actual print and then drops and ruins the print.
- I'm also trying to do a heat soak, because I have a very thick bed (9mm) and it takes a long time for the top surface to get even close to temp. (it never actually gets fully to temp on top) And leaving the hotend heated and oozing for many minutes would cause a huge mess.

On my old X5SA (using an out-of-date version of klipper-macros) I'm using this to set the bed heat soak wait time:
```
#variable_start_bed_heat_delay: 480000  #(8 minutes)
variable_start_bed_heat_delay: 180000  #(3 minutes)
```

When I change filament from PLA to ABS/PETG, I comment out the 3 minute wait and uncomment the 8 minute wait. Not automatic at all, but it works. In an ideal world, I want to scale the wait based on the bed temp with some sort of formula, but I need to do some testing with a thermistor on top to figure out accurate wait times. (but for example: 60C = 3min soak, 90C = 8min soak, etc) And then obviously figure out how to code that into a start macro.

For now I've just switched back to a manually created print_start macro. I have a Beacon probe on backorder, so it sounds like once I add that to the system and take a bed mesh before the print, the hotend will wait to heat until the bed stabilizes and the probing is done using klipper-macros, right?

I do still think if it's not too much trouble, it'd be great to add at least an option for separate bed/extruder heating phases without a probe. But I'd understand if you decide not to. Thanks for making such a great macro system!

**Comment by jschuh** [comment]
After pondering it for a bit, I landed c7d192fd6232341a9390b84235f6cc68ef5450a5 on dev. Would that address your need?

**Reply by Lincoln-Jones** [reply]
I think so? If I'm reading the code correctly, setting that `km.start_extruder_probing_temp` value to 0 would make it warm to the regular extruder temp immediately, right? So I would want to set it to some value under room temp, and then the script would set that as the extruder's target which would prevent the extruder from heating until the bed was up to temp?

**Reply by jschuh** [reply]
Well, you'd want to set it to something a bit above the ambient temperature of its environment, otherwise Klipper will error out on failing to hit the target temp. Something like e.g. 80 should be totally safe, because it's well below the melting point of pretty much any filament, and safely above the zone where you'd hit an error.

**Reply by jzhvymetal** [reply]
Just my opinion, but a faster way is as follows:

1. Preheat both at given percentages, let's say 90% for the bed and 70% for the extruder in the initial phase with no wait
2. Home all Axis
3. Heat Bed with wait at initial preheat temperature
4. Once the bed reaches the initial preheat temperature, heat both the bed and extruder to 100% values with no wait, and start probing.
5. After probing, Heat Extruder 100% no wait and Bed 100% wait.
6. Heat Extruder 100% wait
7. Start printing.

The reasoning behind this process is that if the bed is going to warp due to heat in the initial phase of heating, this saves time by not requiring the bed to fully stabilize. Instead, utilize the additional probing time to heat both the bed and extruder.

The above can all be canceled if the non-blocking heating script that I shared in a previous post is used. I employ a customized version of this script for my startup scripting, but it requires the use some of the kippler-macros and custom scripting.  I have not had time to merge the changes in the kippler-macros but below it what I use.
 

> ### heatsoak.cfg:
> https://github.com/garethky/klipper-voron2.4-config/blob/mainline/printer_data/config/heatsoak.readme.md
> https://github.com/garethky/klipper-voron2.4-config/blob/mainline/printer_data/config/heatsoak.cfg
> 
> ### Printer.cfg:
> [include heatsoak.cfg]
> 
> [gcode_macro CANCEL_WARMUP]
> gcode:
>   STOP_HEAT_SOAK
>    { action_respond_info( "" )}
> 
> [gcode_macro BED_WARMUP]
> gcode:
>     {% set PER = params.PER | default(100) | float %}
>     {% set TEMP = params.TEMP | default(60) | float %}
>     {% set TEMP = TEMP * PER/100 %}
>     M140 S{TEMP} ;Start heating bed
>     {% if printer.toolhead.homed_axes != "xyz" %}
>         G28 ; Home all axes
>     {% endif %}
>     HEAT_SOAK HEATER='heater_bed' TARGET={TEMP} SOAKER='heater_bed' RATE=2 RATE_SMOOTH=10 HEATING_REPORT_INTERVAL=5
> 
> [gcode_macro EXT_WARMUP]
> gcode:
>     {% set PER = params.PER | default(100) | float %}
>     {% set TEMP = params.TEMP | default(190) | float %}
>     {% set TEMP = TEMP * PER/100 %}
>     M104 S{TEMP} ;Start heating extruder
>     HEAT_SOAK HEATER='extruder' TARGET={TEMP} SOAKER='extruder' RATE=2 RATE_SMOOTH=10 HEATING_REPORT_INTERVAL=5
> 
> ### Cura Start G-Code:
> M190 S0
> M109 S0
> _PRINT_START_PHASE_INIT EXTRUDER={material_print_temperature_layer_0} BED={material_bed_temperature_layer_0} NOZZLE_SIZE={machine_nozzle_size}
> BED_WARMUP TEMP={material_bed_temperature_layer_0} PER=90
> ; Insert custom gcode here.
> _PRINT_START_PHASE_PREHEAT
> ; Insert custom gcode here.
> _PRINT_START_PHASE_PROBING
> ; Insert custom gcode here.
> EXT_WARMUP TEMP={material_print_temperature_layer_0} PER=100
> _PRINT_START_PHASE_EXTRUDER
> ; Insert custom gcode here.
> _PRINT_START_PHASE_PURGE
